api/lib/flight_job_script_api/subprocess.rb

```diff
@@ -0,0 +1,191 @@
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Script Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Script Service, please visit:
+# https://github.com/openflighthpc/flight-job-script-service
+#==============================================================================
+
+module FlightJobScriptAPI
+  class Subprocess
+    class Result < Struct.new(:stdout, :stderr, :exitstatus, :pid); end
+
+    def initialize(dir:, env:, gid:, logger:, timeout:, uid:)
+      @dir = dir
+      @env = env
+      @gid = gid
+      @logger = logger
+      @timeout = timeout
+      @uid = uid
+    end
+
+    def run(cmd, stdin, &block)
+      @stdout = ""
+      @stderr = ""
+      @read_threads = []
+      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      create_pipes
+      run_fork do
+        run_command(cmd, &block)
+      end
+      start_read_threads
+      write_stdin(stdin)
+      status = wait_for_process
+      wait_for_read_threads(start_time)
+      exit_code = determine_exit_code(status)
+
+      Result.new(@stdout, @stderr, exit_code, @pid)
+    ensure
+      close_pipes
+    end
+
+    private
+
+    def run_fork(&block)
+      @pid = Kernel.fork do
+        @out_read.close
+        @err_read.close
+        @in_write.close
+
+        Process::Sys.setgid(@gid)
+        Process::Sys.setuid(@uid)
+        Process.setsid
+
+        block.call if block
+      end
+      @logger.debug("Forked process #{@pid.inspect}")
+      @out_write.close
+      @err_write.close
+      @in_read.close
+    end
+
+    def run_command(cmd, &block)
+      block.call(@out_write, @err_write) if block
+      if cmd.nil?
+        @logger.debug("Nothing to exec")
+        return 
+      end
+
+      opts = {
+        unsetenv_others: true,
+        close_others: true,
+        chdir: @dir,
+        out: @out_write,
+        err: @err_write,
+        in: @in_read,
+      }
+      @logger.debug("Execing cmd")
+      Kernel.exec(@env, *cmd, **opts)
+    end
+
+    def start_read_threads
+      [ [@out_read, @stdout], [@err_read, @stderr] ].each do |io, buffer|
+        @read_threads << Thread.new do
+          begin
+            loop { buffer << io.readpartial(1024) }
+          rescue IOError
+            # NOOP - Both EOF and IO closed errors need to be caught
+          end
+        end
+      end
+    end
+
+    def write_stdin(stdin)
+      return if stdin.nil?
+      # We assume here that stdin is small enough to not cause an issue.  The
+      # process and the read threads have been started so we should be OK with
+      # that assumption.
+      @in_write.write(stdin)
```

Comment:
I think that we're OK with the order that `SubProcess` does things in.  The process is started and will presumably be reading from standard in.  The read threads are started so, the process won't block on writing to stdout/stderr.

I guess we could end up writing a large amount of stdin to a process that doesn't read from stdin.  That would `write_stdin` to block before we reach the time out waiting.  I suppose the safest thing to do would be to fire up another thread to prevent that.  I'll reconsider this when `SubProcess` is extracted to a library.